### PR TITLE
Added new widget to log robot errors and warnings

### PR DESCRIPTION
**Pull request by sauravbanna**
### Description

Before, if a robot crashed or had low battery, we would get a popup dialog to inform us of this. This was not ideal as the dialog box stopped the rest of Thunderscope, and it could not be closed without touching the laptop. So instead, errors are now logged to a new widget.

### Testing Done

Thunderscope runs as normal. Tested with some timed crash protos and seems to be working fine

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue

### Screenshot of new changes

image
image
image
image

### CODE REVIEW

File: src/software/jetson_nano/primitive_executor.h
Comment (nimazareian) [inline src/software/jetson_nano/primitive_executor.h:68]: nit: parameter should be added to javadoc

File: src/proto/message_translation/power_frame_msg.hpp
Comment (nimazareian) [inline src/proto/message_translation/power_frame_msg.hpp]: From the look of it this import and the one above are not necessary

File: src/proto/robot_status_msg.proto
Comment (nimazareian) [inline src/proto/robot_status_msg.proto:138]: Can probably make these a `uint8` since it's an integer representing the percentage. Also, could you add some short comment regarding what they represent?

File: src/software/jetson_nano/services/network/network.cpp
Comment (nimazareian) [inline src/software/jetson_nano/services/network/network.cpp:1]: The changes made to network service are actually already in another PR (#2945), which actually makes some other changes as well. So you don't need to include whatever's in that PR already.

File: src/software/thunderscope/robot_diagnostics/error_log_widgets.py
Comment (nimazareian) [inline src/software/thunderscope/robot_diagnostics/error_log_widgets.py:152]: All `__init__` methods in this class seem to be missing their docstring

File: src/software/thunderscope/robot_diagnostics/error_log_widgets.py
Comment (nimazareian) [inline src/software/thunderscope/robot_diagnostics/error_log_widgets.py:22]: I think we should consider making this error widget generalizable enough to also work for error logs from fullsystem and simulator processes where there is not robot id. Possibly instead of a `robot_id` we can have a `title: str` field. The class name can then be simplified to something like `ErrorLogMessageWidget`. 

Ideally we would also update the `RobotLog` protobuf message to be more general `Log` message, since we also use it in fullsystem and simulator. Though let me know if you think these are better suited for a different PR:
- [ ] `RobotLog` protobuf message type should be updated to remove `robot_id` (safely deprecated using `reserved = 1` `reserved "robot_id"`) and replace it with a more general `string source = 7` field. This field can then be filled in `NetworkLoggerSingleton` -> `NetworkSink` to have a value of `R<robot_id>`, similarly `LoggerSingleton` -> `ProtobufSink` can be updated to take in `Blue`/`Yellow`/`Sim` as the source. The protobuf message type could also be renamed to have a more generic name like `Log`.
- [ ] `g3log_widget.py` print the `log.source` instead of `log.robot_id`
- [ ] Add `RobotLog` to `Simulator.setup_proto_unix_io` so logs from the simulator are also received by Thunderscope.

Since this would be a relatively major change of an existing feature, @itsarune what are your thoughts?

File: src/software/thunderscope/robot_diagnostics/error_log_widgets.py
Comment (nimazareian) [inline src/software/thunderscope/robot_diagnostics/error_log_widgets.py]: nit: to be it seemed slightly confusing by the 6 digit number represented. maybe using a period could help?!
```suggestion
        self.timestamp_label = QLabel(self.timestamp.strftime("%H:%M:%S.%f"))
```

File: src/software/thunderscope/robot_diagnostics/robot_info.py
Comment (nimazareian) [inline src/software/thunderscope/robot_diagnostics/robot_info.py]: Seems to be left over from debugging

File: src/software/thunderscope/robot_diagnostics/robot_info.py
Comment (nimazareian) [inline src/software/thunderscope/robot_diagnostics/robot_info.py]: this line doesn't seem to be doing anything

File: src/software/thunderscope/robot_diagnostics/robot_info.py
Comment (nimazareian) [inline src/software/thunderscope/robot_diagnostics/robot_info.py:350]: Is it possible to color either the text or background of these packet loss indicators? One approach would be to have the color fully transparent for the range [0,10], and from 10 - 30% I would linearly increase the opacity, and after 30%+ it would be fully colored. This way the packet loss could stand out more the higher it is, but also not noticeable for lower packet loss percentages (<10%).

File: src/software/thunderscope/robot_diagnostics/robot_error_log.py
Comment (nimazareian) [inline src/software/thunderscope/robot_diagnostics/robot_error_log.py]: This is helpful for now, but should be removed before the PR is merged

File: src/software/thunderscope/widget_setup_functions.py
Comment (nimazareian) [inline src/software/thunderscope/widget_setup_functions.py:256]: `RobotLog` should also be registered here as well.

File: src/software/thunderscope/robot_diagnostics/error_log_widgets.py
Comment (nimazareian) [inline src/software/thunderscope/robot_diagnostics/error_log_widgets.py]: I think this widget can generally be made more compact so `RobotErrorLog` can be viewed as a smaller overall widget. During an actual match, we layout Thunderscope so we show almost every widget simultaneously on the screen, so it's important for the widgets to not take too much screen space.

**Review by nimazareian — COMMENTED** [review]
Just tried it out locally and it looks great! Just had some suggestions.

File: src/software/thunderscope/robot_diagnostics/error_log_widgets.py
Comment (itsarune) [reply src/software/thunderscope/robot_diagnostics/error_log_widgets.py:22]: that's a good idea, I think that'll be an improvement. If @sauravbanna doesn't have too much on his plate, then we can include it in this PR

File: src/software/thunderscope/robot_diagnostics/robot_info.py
Comment (sauravbanna) [reply src/software/thunderscope/robot_diagnostics/robot_info.py]: oh yeah, I didn't make this change so must've been from the robocup branch

File: src/software/thunderscope/robot_diagnostics/error_log_widgets.py
Comment (sauravbanna) [reply src/software/thunderscope/robot_diagnostics/error_log_widgets.py:152]: done

File: src/software/jetson_nano/services/network/network.cpp
Comment (sauravbanna) [reply src/software/jetson_nano/services/network/network.cpp:1]: sure, i've removed those changes

File: src/proto/robot_status_msg.proto
Comment (sauravbanna) [reply src/proto/robot_status_msg.proto:138]: apparently uint_8 is not a thing in protobuf types for scalars, as far as I can tell

File: src/software/thunderscope/robot_diagnostics/robot_info.py
Comment (sauravbanna) [reply src/software/thunderscope/robot_diagnostics/robot_info.py:350]: i'll do this in a seperate PR, it'll be small so probably easier to review and merge

File: src/proto/robot_status_msg.proto
Comment (nimazareian) [reply src/proto/robot_status_msg.proto:138]: oh ok, nvm then. Didn't realize there is no uint8. We can keep it as is then

**Review by nimazareian — DISMISSED** [review]
:100:

File: src/software/jetson_nano/services/network/network.cpp
Comment (itsarune) [inline src/software/jetson_nano/services/network/network.cpp:22]: I updated this section of the code in this PR: https://github.com/UBC-Thunderbots/Software/pull/2945

If your change depends on this code, merge that branch into this one or omit this change

Comment (nimazareian) [reply src/software/jetson_nano/services/network/network.cpp:22]: @sauravbanna #2945 is now merged, so you should be good to make your changes on top of master.

File: src/software/thunderscope/robot_diagnostics/robot_error_log.py
Comment (itsarune) [inline src/software/thunderscope/robot_diagnostics/robot_error_log.py:200]: Maybe I am understanding this logic wrong but would this mean that if two robots had low batteries, then only one error log would show up?

Comment (sauravbanna) [reply src/software/thunderscope/robot_diagnostics/robot_error_log.py:200]: good catch, fixed it now

File: src/software/thunderscope/robot_diagnostics/robot_error_log.py
Comment (itsarune) [inline src/software/thunderscope/robot_diagnostics/robot_error_log.py:156]: Do we really need two separate timeout constants?

This PR merged ROBOT_FATAL and ROBOT_CRASH:
https://github.com/UBC-Thunderbots/Software/pull/3038

File: src/software/thunderscope/robot_diagnostics/robot_error_log.py
Comment (itsarune) [inline src/software/thunderscope/robot_diagnostics/robot_error_log.py:237]: Is there a performance impact if we keep adding more error logs over time?

Does pyqt handle clearing memory associated with widgets that aren't being displayed anymore or does the user have to do it?

Comment (sauravbanna) [reply src/software/thunderscope/robot_diagnostics/robot_error_log.py:237]: right now they're not cleared, i can add a user close button and also clear them after a certain time

File: src/software/python_bindings.cpp
Comment (itsarune) [inline src/software/python_bindings.cpp]: seems unused in this PR

**Review by itsarune — COMMENTED** [review]
close, just a question

**Review by itsarune — COMMENTED** [review]
LGTM pending merge conflicts

File: src/software/thunderscope/thunderscope_config.py
Comment (itsarune) [inline src/software/thunderscope/thunderscope_config.py:239]: this gets launched with thunderscope_main (no args for communication with robots). Would this make sense? @sauravbanna

Comment (nimazareian) [reply src/software/thunderscope/thunderscope_config.py:239]: I think long term this makes sense. It would be nice to have key error messages show up (e.g. LOG(FATAL) in AI or new protobuf message for specifying unexpected scenarios in AI...).

**Review by nimazareian — APPROVED** [review]

**Comment by nimazareian** [comment]
@sauravbanna, @itsarune noticed that there is an error log message that pops up by default when we launch AI vs AI. Do you know where this is coming from? Seems like each fullsystem sends out an empty `RobotStatus` proto.
image

If you think it's unrelated to your PR, then we can create an issue for this to get fixed at a later point

**Comment by nimazareian** [comment]
Actually it seems unrelated to your changes. The simulator is constantly sending `RobotStatus` messages

https://github.com/UBC-Thunderbots/Software/blob/e4f5837793c84c080c99d88ce987d1e1cc10323a/src/software/er_force_simulator_main.cpp#L197-L205

**Review by itsarune — APPROVED** [review]
Thanks for the patience, looks awesome!

Two notes:
1. er_force_simulator_main is spending robot status messages in Ai vs AI that's giving us false robot logs in AI vs AI
2. in the future, it would be great to migrate the error codes that we have in the robot info widget about motor faults to this new logger widget

I feel that both of these two tasks could be tackled in a separate issue, so @sauravbanna would you add issues for these?

**Pull request merged**
